src/main/java/org/generationcp/middleware/api/brapi/v2/observationunit/ObservationUnitServiceImpl.java

```diff
@@ -38,16 +89,302 @@ public void update(final String observationUnitDbId, final ObservationUnitPatchR
 			LOG.error(message, e);
 			throw new MiddlewareException(message);
 		}
 	}
+
+	@Override
+	public List<ObservationUnitDto> searchObservationUnits(final Integer pageSize, final Integer pageNumber,
+		final ObservationUnitSearchRequestDTO requestDTO) {
+		final List<ObservationUnitDto> dtos = this.daoFactory.getPhenotypeDAO().searchObservationUnits(pageSize, pageNumber, requestDTO);
+		if(!CollectionUtils.isEmpty(dtos)) {
+			final List<Integer> experimentIds = dtos.stream().map(o -> Integer.valueOf(o.getExperimentId())).collect(Collectors.toList());
+
+			final Map<String, List<ExternalReferenceDTO>> externalReferencesMap =
+				this.daoFactory.getExperimentExternalReferenceDao().getExternalReferences(experimentIds).stream()
+					.collect(groupingBy(
+						ExternalReferenceDTO::getEntityId));
+
+			final Map<Integer, List<ObservationLevelRelationship>> obsevationRelationshipsMap =
+				this.daoFactory.getExperimentPropertyDao().getObservationLevelRelationshipMap(experimentIds);
+
+			for (final ObservationUnitDto dto : dtos) {
+				dto.setExternalReferences(externalReferencesMap.get(dto.getExperimentId().toString()));
+				dto.getObservationUnitPosition().setObservationLevelRelationships(obsevationRelationshipsMap.get(dto.getExperimentId()));
+			}
+		}
+
+		return dtos;
+	}
+
+	@Override
+	public long countObservationUnits(final ObservationUnitSearchRequestDTO requestDTO) {
+		return this.daoFactory.getPhenotypeDAO().countObservationUnits(requestDTO);
+	}
+
+	@Override
+	public List<ObservationUnitDto> importObservationUnits(final String crop, final List<ObservationUnitImportRequestDto> requestDtos) {
+		final CropType cropType = this.daoFactory.getCropTypeDAO().getByName(crop);
+
+		final List<String> germplasmDbIds =
+			requestDtos.stream().map(ObservationUnitImportRequestDto::getGermplasmDbId).collect(Collectors.toList());
+		final GermplasmSearchRequestDto germplasmSearchRequestDto = new GermplasmSearchRequestDto();
+		germplasmSearchRequestDto.setGermplasmDbIds(germplasmDbIds);
+		final Map<String, GermplasmDTO> germplasmDTOMap = this.germplasmService.searchFilteredGermplasm(germplasmSearchRequestDto, null)
+			.stream().collect(Collectors.toMap(GermplasmDTO::getGermplasmDbId, Function.identity()));
+
+		final List<Integer> trialIds = requestDtos.stream().map(r -> Integer.valueOf(r.getTrialDbId())).collect(Collectors.toList());
+		final List<String> variableNames = new ArrayList<>();
+		requestDtos.stream().forEach(dto -> {
+			if (!CollectionUtils.isEmpty(dto.getObservationUnitPosition().getObservationLevelRelationships())) {
+				variableNames.addAll(dto.getObservationUnitPosition().getObservationLevelRelationships().stream()
+					.map(ObservationLevelRelationship::getLevelName).collect(Collectors.toList()));
+			}
+		});
+
+		final Map<Integer, List<StockModel>> stockMap = this.daoFactory.getStockDao().getStockMapByStudyIds(trialIds);
```

Comment:
Refactor getStockMapByStudyIds to getStocksByStudyIds and return a list instead. Then just convert it to map outside of dao method. You can also transform the list to <Map, Map<String, StockModel>> so that it's easier to lookup the stock per germplasmDbId.